### PR TITLE
Nav group state

### PR DESCRIPTION
**Pull request by errael**
Refer to #40 for discussion. In particular the 4th post that starts
> This post refers in particular to the NavGroupSate PR 

In SSDataNavigator lambdas instead of anonymous classes (which also gets back some indent). So turning on ignore white space in the diffs is recommended.

If this is merged, there should be no visible changes. The above post describes the properties for turning on the new features.  The discussion mentions possible us of the infrastructure `EventBus` introduced in the PR for v4.

### CODE REVIEW

File: swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java [hunk 1]
```diff
@@ -363,6 +398,50 @@ public SSDataNavigator(final RowSet _rowSet, final Dimension _buttonSize) {
...
+			eventBus = getLocalEventBus(this, rowSet);
+		}
+		eventBus.register(new BusReceiver());
+	}
+
+	// TODO: also have Set<SSComponentInterface> modifiedComponents
+	private final Set<SSComponentInterface> errorComponents = new HashSet<>();
+	class BusReceiver {
+		@Subscribe
+		public void handleRowDataChanged(RowSetModificationEvent ev) {
```
Comment (errael) [inline swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java:419]: This is how to listen to the event bus

File: swingset/src/main/java/com/nqadmin/swingset/utils/SSUtils.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+	/**
+	 * Post an error modification event; prefer a local eventBus.
+	 * @param source SSComponent modifying the rowset
+	 * @param _rowSet The modified row set
+	 */
+	public static void postRowSetModifiedError(SSComponentInterface source, RowSet _rowSet) {
+		getLocalEventBus(source, _rowSet)
+				.post(new RowSetModificationEvent(source, _rowSet, true));
+	}
```
Comment (errael) [inline swingset/src/main/java/com/nqadmin/swingset/utils/SSUtils.java:119]: Broadcasting an event over the bus

File: swingset/src/main/java/com/nqadmin/swingset/utils/CentralLookup.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Class used to house anything one might want to store
+ * in a central lookup which can affect anything within
+ * the application. It can be thought of as a central context
+ * where any application data may be stored and watched.
+ * 
+ * A singleton instance is created using @see getDefault().
+ * This class is as thread safe as Lookup. Lookup appears to be safe.
+ * @author Wade Chandler
```
Comment (bpangburn) [inline swingset/src/main/java/com/nqadmin/swingset/utils/CentralLookup.java:15]: An alias?

File: swingset/pom.xml [hunk 4]
```diff
@@ -96,6 +98,21 @@
 			<artifactId>jdbcrowsetimpl</artifactId>
 			<version>${version.jdbcrowsetimpl}</version>
 		</dependency>
+
+		<!-- https://mvnrepository.com/artifact/org.netbeans.api/org-openide-util-lookup -->
+		<dependency>
+			<groupId>org.netbeans.api</groupId>
+			<artifactId>org-openide-util-lookup</artifactId>
+			<version>${version.netbeans}</version>
+		</dependency>
```
Comment (bpangburn) [inline swingset/pom.xml:107]: I see this is for the CentralLookup class. Mentioned for some boolean constants, but it appears it is used to store state data for Event Bus?

File: swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java [hunk 5]
```diff
@@ -329,12 +352,22 @@ private void performUpdates() {
...
+	//
+	// TODO:
+	//     For now, have the defaults here. In the future,
+	//     probably want to set the defaults from some
+	//     configurable spot, via CentralLookup?
+	//     Maybe: interface SwingSetConfiguration {}
+	//
+	private static final boolean V3_BUTTONS_DEFAULT = true;
+	private static final boolean AUTO_COMMIT_DEFAULT = false;
+
```
Comment (bpangburn) [inline swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java:364]: We have swingset.properties in src/main/resources too.

File: swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java [hunk 6]
```diff
@@ -1350,6 +1349,147 @@ public void setRowSet(final RowSet _rowSet) {
...
 //	}
 
+	private void setRowModified(boolean isDirty) {
+		isRowModified = isDirty;
+		if (!isDirty) {
+			errorComponents.clear();
+		}
+	}
+
+	private void updb(JButton b, boolean flag) {
```
Comment (bpangburn) [inline swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java:1359]: What is updb doing?

File: swingset/pom.xml (hunk 4 above)
Comment (errael) [reply swingset/pom.xml:107]: As discussed elsewhere, this is a way to make stuff accessible without an API. For the event bus, there is an API in SSUtils.

File: swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java (hunk 5 above)
Comment (errael) [reply swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java:364]: This is a placeholder, and a way to indicate that this is something different screens/apps/whatever might want to configure globally.

File: swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java (hunk 6 above)
Comment (errael) [reply swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java:1359]: updateButtonEnableState. It's a minimal overhead way to set a buttons state. It's used a lot. (but it probably doesn't matter overall performance); it is a one liner.

**Pull request closed without merging**